File: src/main/java/com/plugin/frege/documentation/FregeDocumentationProvider.java
package com.plugin.frege.documentation;

import com.intellij.lang.documentation.AbstractDocumentationProvider;
import com.intellij.openapi.util.NlsSafe;
import com.intellij.psi.PsiElement;
import com.plugin.frege.psi.FregeDocumentationElement;
import org.jetbrains.annotations.Nullable;

import java.util.stream.Collectors;

public class FregeDocumentationProvider extends AbstractDocumentationProvider {
    @Override
    public @Nullable String getQuickNavigateInfo(PsiElement element, PsiElement originalElement) {
        return null; // TODO
    }

    @SuppressWarnings("UnstableApiUsage")
    @Override
    public @Nullable @NlsSafe String generateDoc(PsiElement element, @Nullable PsiElement originalElement) {
        if (!(element instanceof FregeElementProvideDocumentation)) {
            return null;
        }
        FregeElementProvideDocumentation elementProvideDocumentation = (FregeElementProvideDocumentation) (element);
        return elementProvideDocumentation.getDocs().stream().map(FregeDocumentationElement::getDocumentationText).collect(Collectors.joining("\n"));
    }
}

File: src/main/kotlin/com/plugin/frege/documentation/FregeDocumentationProvider.kt
package com.plugin.frege.documentation

import com.intellij.codeInsight.javadoc.JavaDocUtil
import com.intellij.lang.documentation.AbstractDocumentationProvider
import com.intellij.openapi.project.DumbService
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiManager
import com.plugin.frege.psi.FregeDocumentableElement
import com.plugin.frege.psi.impl.FregeBindingImpl

class FregeDocumentationProvider : AbstractDocumentationProvider() {
    override fun generateDoc(element: PsiElement, originalElement: PsiElement?): String? {
        if (DumbService.isDumb(element.project)) {
            return null
        }
        if (element is FregeBindingImpl) {
            element.getAnnoItem()?.let { return it.generateDoc() }
        }
        return (element as? FregeDocumentableElement)?.generateDoc()
    }

    override fun getDocumentationElementForLink(
        manager: PsiManager?,
        link: String?,
        context: PsiElement?
    ): PsiElement? {
        if (manager == null || link == null) {
            return null
        }
        return JavaDocUtil.findReferenceTarget(manager, link, context, false)
    }
}
